RNWebView: (inpage provider) Replace script tag with straight code execution

jpcloureiro opened the pull request:
<!--
Please submit this PR as a draft initially.
Do not mark it as "Ready for review" until the template has been completely filled out, and PR status checks have passed at least once.
-->

## **Description**

CSP directives might block inline script execution, thus preventing our web3 provider from being injected with a script tag creation.

Replacing it with a direct function call avoids being blocked. 

`inpage-bridge` build process updated to include initial validations prior to inpage-content execution


## **Related issues**

Fixes: https://github.com/MetaMask/mobile-planning/issues/1618#issuecomment-2147362799

## **Manual testing steps**

1. Go to this page...
2.
3.

## **Screenshots/Recordings**

<!-- If applicable, add screenshots and/or recordings to visualize the before and after of your change. -->

### **Before**

<!-- [screenshots/recordings] -->

### **After**

<!-- [screenshots/recordings] -->

## **Pre-merge author checklist**

- [ ] I’ve followed [MetaMask Contributor Docs](https://github.com/MetaMask/contributor-docs) and [MetaMask Mobile Coding Standards](https://github.com/MetaMask/metamask-mobile/blob/main/.github/guidelines/CODING_GUIDELINES.md).
- [ ] I've completed the PR template to the best of my ability
- [ ] I’ve included tests if applicable
- [ ] I’ve documented my code using [JSDoc](https://jsdoc.app/) format if applicable
- [ ] I’ve applied the right labels on the PR (see [labeling guidelines](https://github.com/MetaMask/metamask-mobile/blob/main/.github/guidelines/LABELING_GUIDELINES.md)). Not required for external contributors.

## **Pre-merge reviewer checklist**

- [ ] I've manually tested the PR (e.g. pull and build branch, run the app, test code being changed).
- [ ] I confirm that this PR addresses all acceptance criteria described in the ticket it closes and includes the necessary testing evidence such as recordings and or screenshots.

jpcloureiro committed:
test inpage content

Cal-L commented:
@jpcloureiro Nice job getting this to work! Could you provide context into the change in underlying logic here and what makes this solution work?

jpcloureiro committed:
remove inpage content injection

Remove inline code injection and doc element validation

Inpage content must be executed through native webview JS evaluator

jpcloureiro committed:
inpage-content: export to separate file

jpcloureiro committed:
copy inpage content file to android assets folder

jpcloureiro committed:
InpageContentWeb3: load inpage-content.js

Load Inpage Content and pass it through Webview injectedJS prop

jpcloureiro committed:
update gitignore: add InpageContent assets file

Cal-L commented on scripts/inpage-bridge/content-script/index.js:
```diff
@@ -47,7 +24,6 @@ function shouldInject() {
   return (
     doctypeCheck() &&
     suffixCheck() &&
-    documentElementCheck() &&
```
Do we need to remove this for the new injection to work?

jpcloureiro replied:
Yes, otherwise provider would fail to init with the following error

<img width="1332" alt="image" src="https://github.com/MetaMask/mobile-planning/assets/1649425/27ddefc5-ae7a-4ab9-9e44-0a87d264f619">

jpcloureiro commented:
Loading two JS files from the filesystem & passing them through the RN bridge to the native webview is too heavy and affects the user experience too much.

Reverting this portion of the code [code changes that will be reverted](https://github.com/MetaMask/metamask-mobile/pull/10053/files/3c942617b9895d7ce093cbd8539c244c3bff8b84..4f11562f49d868f06b131cfd04b303dcb808dba1)

jpcloureiro commented:
Loading two JS files from the filesystem & passing them through the RN bridge to the native webview is too heavy and affects the user experience too much.

[Reverting this portion of the code](https://github.com/MetaMask/metamask-mobile/pull/10053/files/3c942617b9895d7ce093cbd8539c244c3bff8b84..4f11562f49d868f06b131cfd04b303dcb808dba1)

jpcloureiro committed:
Revert up to 3c94261

jpcloureiro committed:
test inline script

jpcloureiro committed:
move inline script to browser scripts file

jpcloureiro committed:
remove dangling file

jpcloureiro committed:
fix lint

jpcloureiro committed:
code origin reference

jpcloureiro committed:
Merge branch 'upgrade/react-native-webview' into rnwebview-web3provider-injection

jpcloureiro committed:
remove blank lines

jpcloureiro committed:
change inpage-bridge build process

remove browser script string with duplicated code.

Change inpage-bridge bundle to build a single index.js
with both inpage provider and content-provider validations.

jpcloureiro committed:
Remove documentElementCheck

Avoid TypeError: Cannot read properties of null (reading 'nodeName')

The script excutes fine without documentElement.

jpcloureiro committed:
Merge branch 'upgrade/react-native-webview' into rnwebview-web3provider-injection

tommasini reviewed (approved):
AWESOME!!!!!!

jiexi commented on scripts/inpage-bridge/src/index.js:
```diff
@@ -47,7 +39,6 @@ function shouldInject() {
   return (
     doctypeCheck() &&
     suffixCheck() &&
-    documentElementCheck() &&
```
still unsure how important this check is. @Gudahtt

jiexi replied:
@jpcloureiro were we planning on performing this check outside of the WebView component? or did we determine that that isn't an option either?

jpcloureiro replied:
There are plans to optimize the flow in which the provider is injected. These checks will undergo the same optimization.

For now we'll keep it in the Webview component context

jiexi commented on scripts/inpage-bridge/src/index.js:
```diff
@@ -21,18 +22,9 @@ async function start() {
  * @param {string} content - Code to be executed in the current document
  */
-function injectScript(content) {
+function injectScript() {
```
can probably inline this function into line 5 since it doesn't really do much anymore.

Not even sure if the try catch really brings any value

jiexi commented on scripts/inpage-bridge/src/index.js:
```diff
@@ -1,7 +1,8 @@
-/* global inpageBundle */
+
+import inpageProvider from './provider';
```
nit. do you have thoughts on changing this name to be a little more descriptive? 

```suggestion
import initializeInpageProvider from './provider';
```

jpcloureiro replied:
Since we're adding new properties to the `window` object, does `injectInpageProvider` sound more accurate?

jiexi replied:
works for me

jpcloureiro committed:
Update default import name

Co-authored-by: jiexi <jiexiluan@gmail.com>

jpcloureiro commented:
Development will continue on the main feature branch `upgrade/react-native-webview
